### PR TITLE
Implementation of transmit method

### LINKED ISSUE #26: Read data from WiFi queue and transmit over LoRa mesh

### PR DESCRIPTION
**Pull request by IvoAP**
We implemented the method transmit over loramesh. closes #26.

### CODE REVIEW

File: src/lorarmesh.cpp [hunk 1]
```diff
@@ -14,38 +30,93 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
     free(message->data.layer2.payload);
   }
   free(message);
 }
 
 void LoraMesh::receive() {
-  message_t* message = nullptr;  // TODO read actual data from LoRaLayer2
-  if (message == nullptr)
-    return;
+  message_t* message = nullptr;
```
Comment (alexfrt) [inline src/lorarmesh.cpp:96]: Why are you changing the receive() method if the goal is to transmit? This should be done in other PR.

**Review by alexfrt — CHANGES_REQUESTED** [review]
Why do we need this Teste/teste.cpp file? Please delete it.

File: src/lorarmesh.cpp [hunk 2]
```diff
@@ -1,4 +1,19 @@
...
+#define LORA_RST 14
+#define LORA_IRQ 26
+#define LORA_FREQ 915E6
+#define LED 25
+#define TX_POWER 20
+
+#define DATAGRAM_HEADER 5
+#define MESSAGE_LENGTH 233
+
+const int BROADCAST_NODES[ADDR_LENGTH] = {0xff, 0xff, 0xff, 0xff};
```
Comment (alexfrt) [inline src/lorarmesh.cpp:16]: Should it be static?

File: src/lorarmesh.cpp [hunk 3]
```diff
@@ -14,38 +29,71 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
+  {
+    newTry++;
+  }
+  if (newTry < 30){
+    LL2 = new LL2Class(Layer1);  
+    LL2->setInterval(10000); // set to zero to disable routing packets
+    LL2->init();
+  }else{
+    delay(500);
+    ESP.restart();
```
Comment (alexfrt) [inline src/lorarmesh.cpp:48]: Take a look on how this is usually done ([Click here](https://github.com/Instituto-Atlantico/mesh-connect/blob/master/src/gateway.cpp#L45))

File: src/lorarmesh.cpp [hunk 4]
```diff
@@ -14,38 +29,71 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
-freeData:
+  
   if (message->type == DATA_MESSAGE) {
     free(message->data.layer2.payload);
   }
   free(message);
 }
 
 void LoraMesh::receive() {
-  message_t* message = nullptr;  // TODO read actual data from LoRaLayer2
```
Comment (alexfrt) [inline src/lorarmesh.cpp:44]: Do not change the receive method.

**Review by alexfrt — CHANGES_REQUESTED** [review]

**Comment by alexfrt** [comment]
Please fix the conflicting files @IvoAP @PietroAndrade

**Review by alexfrt — CHANGES_REQUESTED** [review]
Please merge the conflicting files and address all previous requested changes.

File: src/lorarmesh.cpp (hunk 1 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp]: Done

File: src/lorarmesh.cpp (hunk 3 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:47]: it's done, but i implemented this way

Comment (PietroAndrade) [reply src/lorarmesh.cpp:47]: while (!Layer1->init()) {
    delay(CONNECT_DELAY_MILLS);
    if (attempts-- == 0)
      ESP.restart();
  }

File: src/lorarmesh.cpp (hunk 2 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:16]: Yeah, it's done!

File: src/lorarmesh.cpp (hunk 4 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:44]: `void LoraMesh::receive() {
  message_t* message = nullptr;  // TODO read actual data from LoRaLayer2
  if (message == nullptr)
    return;

  if (message->type == CONTROL_MESSAGE) {
    router->handleControlMessage(message);
  }
}`

File: include/loramesh.h [hunk 5]
```diff
@@ -14,6 +16,10 @@ class LoraMesh {
   DataQueue<message_t>* txQueue;
   DataQueue<message_t>* rxQueue;
   Router* router;
+  Layer1Class* Layer1;
+  LL2Class* LL2;
+  char routes[256];
+  char neighbors[256];
```
Comment (alexfrt) [inline include/loramesh.h:22]: Both routes and neighbors are not used, thus not necessary. Please remove it.

File: src/lorarmesh.cpp [hunk 6]
```diff
@@ -1,5 +1,22 @@
+#include <Layer1_LoRa.h>
+#include <LoRaLayer2.h>
 #include "loramesh.h"
 
+#define ADDR_LENGTH 4
+#define LORA_CS 18
+#define LORA_RST 14
+#define LORA_IRQ 26
+#define LORA_FREQ 915E6
+#define LED 25
```
Comment (alexfrt) [inline src/lorarmesh.cpp:10]: Not used. Please remove.

File: src/lorarmesh.cpp [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+#define TX_POWER 20
+
+#define DATAGRAM_HEADER 5
+#define MESSAGE_LENGTH 233
```
Comment (alexfrt) [inline src/lorarmesh.cpp:14]: Not used. Please remove.

File: src/lorarmesh.cpp [hunk 8]
```diff
@@ -1,5 +1,22 @@
...
+#define CONNECT_DELAY_MILLS 500
+#define MAX_CONNECT_RETRIES 10
```
Comment (alexfrt) [inline src/lorarmesh.cpp:16]: **Connect** delay and **connect** retries? Layer1->init() doesn't seem to connect anywhere. Please rename it.

File: src/lorarmesh.cpp [hunk 9]
```diff
@@ -14,25 +31,60 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
+
+  int attempts = MAX_CONNECT_RETRIES;
+
+  while (!Layer1->init()) {
+    delay(CONNECT_DELAY_MILLS);
+    if (attempts-- == 0)
+      ESP.restart();
+  }
+  LL2 = new LL2Class(Layer1);
+  LL2->setInterval(10000);  // set to zero to disable routing packets
```
Comment (alexfrt) [inline src/lorarmesh.cpp:48]: Why 10000? What this value means?

File: src/lorarmesh.cpp [hunk 10]
```diff
@@ -14,25 +31,60 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
     auto destinationAddr = router->getGatewayAddress();
-    if (destinationAddr == 0)
-      goto freeData;
+    if (destinationAddr > 0) {
+      datagram.type = message->type;
+      memcpy(datagram.destination, &destinationAddr, ADDR_LENGTH);
+      memcpy(datagram.message, message->data.layer2.payload,
+             message->data.layer2.length);
+      datagramsize = DATAGRAM_HEADER;
+      datagramsize += message->data.layer2.length;
```
Comment (alexfrt) [inline src/lorarmesh.cpp:79]: `datagramsize = DATAGRAM_HEADER` and `datagram.type = message->type` are duplicated on both conditions.

File: src/lorarmesh.cpp [hunk 11]
```diff
@@ -14,25 +31,60 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
 
-    // TODO send to destinationAddr through LoRaLayer2
+    } else {
+      goto freeData;
```
Comment (alexfrt) [inline src/lorarmesh.cpp:82]: Try to not use _goto_.

**Review by alexfrt — CHANGES_REQUESTED** [review]

File: src/lorarmesh.cpp (hunk 3 above)
Comment (alexfrt) [reply src/lorarmesh.cpp]: Nice!

File: src/lorarmesh.cpp (hunk 10 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:79]: essas linhas foram movidas para antes das condições do tipo de 'message'

Comment (PietroAndrade) [reply src/lorarmesh.cpp:79]: `LL2->daemon();`
`  struct Datagram datagram;`
`  int datagramsize = DATAGRAM_HEADER;`
`  datagram.type = message->type;`
 ` if (message->type == CONTROL_MESSAGE) { ... }`
` else if (message->type == DATA_MESSAGE) { ...}`

File: src/lorarmesh.cpp (hunk 9 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:48]: 10000 is the value of the `_routingInterval variable`, which is used as the data sending interval for LoRaLayer2. It's used in the condition for receiving or transmitting data.

Comment (PietroAndrade) [reply src/lorarmesh.cpp:48]: We still don't know which value to use. Research is needed to define an appropriate value.

File: include/loramesh.h (hunk 5 above)
Comment (PietroAndrade) [reply include/loramesh.h:22]: Essas variáveis foram declaradas por um momento que precisamos, mas não precisamos mais delas. Thanks

File: src/lorarmesh.cpp (hunk 9 above)
Comment (alexfrt) [reply src/lorarmesh.cpp:48]: I opened issue #36 to investigate that.

File: src/lorarmesh.cpp (hunk 6 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:10]: This section `#define LED 25` is the only one that isn't being used.

Comment (PietroAndrade) [reply src/lorarmesh.cpp:10]: The others are used to initialize the `Layer1->setPins`

File: src/lorarmesh.cpp (hunk 7 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:14]: MESSAGE_LENGTH was removed for being unusable

File: src/lorarmesh.cpp (hunk 8 above)
Comment (PietroAndrade) [reply src/lorarmesh.cpp:16]: They were replaced by:
`BOOT_LL1_DELAY_MICROS 500`
`MAX_BOOT_LL1_RETRIES 10`

File: include/loramesh.h [hunk 12]
```diff
@@ -14,6 +16,8 @@ class LoraMesh {
   DataQueue<message_t>* txQueue;
   DataQueue<message_t>* rxQueue;
   Router* router;
+  Layer1Class* Layer1;
+  LL2Class* LL2;
```
Comment (alexfrt) [inline include/loramesh.h:20]: Class attributes must be camelCase, with first char lowercase.

File: src/lorarmesh.cpp [hunk 13]
```diff
@@ -14,29 +30,64 @@ LoraMesh::LoraMesh(DataQueue<message_t>* txQueue,
...
+      memcpy(datagram.destination, &destinationAddr, ADDR_LENGTH);
+      memcpy(datagram.message, message->data.layer2.payload,
+             message->data.layer2.length);
+      datagramsize += message->data.layer2.length;
+      checkTransmitMessage = true;
+    }
+  }
 
-    // TODO send to destinationAddr through LoRaLayer2
+  if (checkTransmitMessage) {
```
Comment (alexfrt) [inline src/lorarmesh.cpp:83]: [OPTIONAL] Rename this variable to "shouldTransmitt" or check `if (datagramsize > DATAGRAM_HEADER)` so you don't neeed a separate flag...

**Review by alexfrt — CHANGES_REQUESTED** [review]
Almost there

File: src/lorarmesh.cpp (hunk 11 above)
Comment (IvoAP) [reply src/lorarmesh.cpp]: We use a boolean to check if the data should be write ant we remove the goto.

File: src/lorarmesh.cpp (hunk 13 above)
Comment (IvoAP) [reply src/lorarmesh.cpp]: We rename the variable to shouldtransmit

File: include/loramesh.h (hunk 12 above)
Comment (IvoAP) [reply include/loramesh.h]: We rename layer1 to layer1 and LL2 to ll2

**Review by alexfrt — APPROVED** [review]

**Comment by IvoAP** [comment]
We merge with the mastre the implementation of the method transmit of loramesh.cpp

**Pull request merged**